### PR TITLE
Differential Privacy

### LINKED ISSUE #303: Add differential privacy
image (1)

### PR DESCRIPTION
**Pull request by s314cy**
Closes #303

### CODE REVIEW

**Review by martinjaggi — CHANGES_REQUESTED** [review]
looking good. 2 things
- you're doing it for the weights, not the \delta w (gradients). while for adding noise that can be fine, the clipping unfortunately is not ok on models. that would need to be on gradients. we need to briefly discuss maybe if in FL we want to send gradients or models. without DP, the result in both cases is identical anyway. we might later also want to support byzantine robustness, which also likes to have clipping on gradients, not models. for the moment what you could do is keep sending models, but before you do, the 2 new ops you now added should happen only on the local gradients, not the models
- add a field to the task config where people can switch it on/off andspecify the noise level

**Comment by s314cy** [comment]
> looking good. 2 things
> 
>     * you're doing it for the weights, not the \delta w (gradients). while for adding noise that can be fine, the clipping unfortunately is not ok on models. that would need to be on gradients. we need to briefly discuss maybe if in FL we want to send gradients or models. without DP, the result in both cases is identical anyway. we might later also want to support byzantine robustness, which also likes to have clipping on gradients, not models. for the moment what you could do is keep sending models, but before you do, the 2 new ops you now added should happen only on the local gradients, not the models
> 
>     * add a field to the task config where people can switch it on/off andspecify the noise level

I fully agree. At first glance, there doesn't seem to be a way to access the current epoch's gradient, which is why @Nacho114 and I decided to go with the weights. Would taking the difference W_curr - W_prev be correct? This doesn't take into account the current learning rate.

I'll add the task config field for the noise scale 👍

**Comment by martinjaggi** [comment]
exactly. the diff curr-prev (over the round) is perfectly fine. then clip etc. if we do it like this, the nice thing is that it will work easily for both federated and decentralized

**Comment by martinjaggi** [comment]
actually you can keep sending weights (model) instead of diffs.

it's just that locally before sending, the client needs to compute their own weight diff since beginning of the round, and then apply noise and clipping. the result can still be the weights (model)

maybe easier this way because then it will be the same in decentralized

**Comment by s314cy** [comment]
> actually you can keep sending weights (model) instead of diffs.
> 
> it's just that locally before sending, the client needs to compute their own weight diff since beginning of the round, and then apply noise and clipping. the result can still be the weights (model)
> 
> maybe easier this way because then it will be the same in decentralized

Oh right, I see

File: browser/src/communication/federated/federated_client.ts [hunk 1]
```diff
@@ -119,13 +123,34 @@ export class FederatedClient extends Client {
     }
   }
 
+  private addDifferentialPrivacy (currentRoundWeights: Weights): Weights {
```
Comment (martinjaggi) [inline browser/src/communication/federated/federated_client.ts]: make state-less, by also giving `previousRoundWeights = this.previousRoundWeights` as a param.

File: browser/src/communication/federated/federated_client.ts [hunk 2]
```diff
@@ -119,13 +123,34 @@ export class FederatedClient extends Client {
...
+    const weightsDiff = List(currentRoundWeights)
+      .zip(List(this.previousRoundWeights))
+      .map(([w1, w2]) => w1.add(-w2))
+
+    const normBound = weightsDiff.map((w) => w.norm().dataSync()[0]).max()
```
Comment (martinjaggi) [inline browser/src/communication/federated/federated_client.ts]: add a comment what kind of norm you compute. why is there a max? norm is 2-norm but at the end gets maxed, like infty-norm?

File: browser/src/communication/federated/federated_client.ts [hunk 3]
```diff
@@ -119,13 +123,34 @@ export class FederatedClient extends Client {
...
+
+    const noisyWeightsDiff = weightsDiff.map((w) => {
+      const clipped = w.div(Math.max(1, w.norm().dataSync()[0] / normBound))
```
Comment (martinjaggi) [inline browser/src/communication/federated/federated_client.ts]: the clipping radius (the maximum norm the grad can be after clipping) should be a parameter in the task config as well. if set to infty, there will be no clipping

File: browser/src/communication/federated/federated_client.ts [hunk 4]
```diff
@@ -119,13 +123,34 @@ export class FederatedClient extends Client {
...
+      const noise = tf.randomNormal(w.shape, 0, normBound * normBound)
```
Comment (martinjaggi) [inline browser/src/communication/federated/federated_client.ts]: rename normBound -> norm ?  also make a configurable scalar parameter (from taskconfig) in front of this one? if zero, there will be no noise added (actually pack this one into an if-statement, as it's probably quite slow if there are millions of weights. rand generators can take quite some time.  this way the DP can be switched on / off in the task config (also add a comment to explain this in the config and here)

File: browser/src/communication/federated/federated_client.ts [hunk 5]
```diff
@@ -119,13 +123,34 @@ export class FederatedClient extends Client {
...
+
   private async fetchServerStatisticsAndUpdateInformant (trainingInformant: TrainingInformant) {
     const serverStatistics = await api.getAsyncWeightInformantStatistics(this.task.taskID, this.clientID)
     trainingInformant.updateWithServerStatistics(serverStatistics.data.statistics)
   }
 
   async onRoundEndCommunication (model: tf.LayersModel, _: number, trainingInformant: TrainingInformant) {
-    await this.postWeightsToServer(model.weights.map((w) => w.read()))
+    const weights = model.weights.map((w) => w.read())
+    const noisyWeights = this.addDifferentialPrivacy(weights)
```
Comment (martinjaggi) [inline browser/src/communication/federated/federated_client.ts]: as mentioned, better stateless with 2 params (prev and curr weights).

**Review by martinjaggi — APPROVED** [review]
very nice. only some cosmetic comments.
also did you try run it and observed any effect on accuracy? does it work on titanic and image tasks?

File: browser/src/communication/federated/federated_client.ts (hunk 2 above)
Comment (s314cy) [reply browser/src/communication/federated/federated_client.ts]: that's an error on my side! will push changes soon

File: browser/src/communication/federated/federated_client.ts (hunk 1 above)
Comment (s314cy) [reply browser/src/communication/federated/federated_client.ts]: I agree on the benefits of stateless-ness, however `Client` is the class actually updating the `Trainer`'s model, hence the one which has direct access to both the previous and current model. Moving the state to `Trainer` would require more effort and might not be cleaner 🤔 

EDIT: I'll try a few things

File: browser/src/communication/federated/federated_client.ts (hunk 4 above)
Comment (s314cy) [reply browser/src/communication/federated/federated_client.ts]: do we still want clipping when there is no noise added?

Comment (martinjaggi) [reply browser/src/communication/federated/federated_client.ts]: these can be independent. the default in the task config would be `noiseLevel=0` and `clipping=infty`. so both deactivated. so if people want they can activate them separately (clipping gives robustness for example)

**Comment by s314cy** [comment]
closing this, will open create another branch since @tharvik's latest PR moved many files

**Comment by martinjaggi** [comment]
exactly. the diff curr-prev (over the round) is perfectly fine

**Pull request closed without merging**